spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/CrossProduct.scala

```diff
@@ -0,0 +1,159 @@
+ *
+ * @param propagateBack whether updateGradInput during backward, default: true
+ * @param numTensor (for checking)number of Tensor input Table contains, default: None
+ * @param embeddingSize (for checking)vector length of dot product, default: None
+ */
+class CrossProduct[T: ClassTag](
+  val propagateBack: Boolean = true,
+  val numTensor: Int = 0,
+  val embeddingSize: Int = 0
+)(implicit ev: TensorNumeric[T]) extends AbstractModule[Table, Tensor[T], T] {
+
+  override def updateOutput(input: Table): Tensor[T] = {
+    val len = input.length()
+    require(numTensor <= 0 || numTensor == len,
+      s"len of input Table($len) not equal to numTensor($numTensor)!")
+
+    val (_, batch, _) = getShape(input[Tensor[T]](1))
+    output.resize(batch, len * (len - 1) / 2)
+
+    if (embeddingSize > 0) {
+      (1 to len).foreach(i => checkEmbeddingSize(input(i)))
+    }
+
+    var cc = 1
+    for( i <- 1 until len; j <- i + 1 to len ) {
+      val ijDot = dot(input(i), input(j))
+      output.select(2, cc).copy(ijDot)
+      cc += 1
+    }
+
+    output
+  }
+
+  override def updateGradInput(input: Table, gradOutput: Tensor[T]): Table = {
+    gradInput = T()
+
+    if (propagateBack) {
+      val (len, gout) = input.length() -> gradOutput
+      require(gout.dim() == 2, s"invalid dim of gradOutput(${gout.dim()})!")
+
+      val outLen = len * (len - 1) / 2
+      require(gout.size(2) == outLen,
+        s"invalid colSize of gradOutput(${gout.size(2)}), it should be $outLen!")
+
+      val (dim, _, emLen) = getShape(input[Tensor[T]](1))
+
+      var cc = 1
+      for( i <- 1 until len; j <- i + 1 to len ) {
+        val (ti, tj) = dim match {
+          case 1 =>
+            input[Tensor[T]](i).view(1, emLen) -> input[Tensor[T]](j).view(1, emLen)
+          case 2 =>
+            input[Tensor[T]](i) -> input[Tensor[T]](j)
+        }
+
+        // get cc_th column data from total gradOut
+        val go = gout.narrow(2, cc, 1)
+
+        val jInc = Tensor[T]().resizeAs(ti).copy(ti).cmul(go)
+        if (dim == 1) jInc.resize(Array(jInc.size(2)))
+        gradInput.get[Tensor[T]](j) match {
+          case None => gradInput.update(j, jInc)
+          case Some(v) => v.add(jInc)
+        }
+
+        val iInc = Tensor[T]().resizeAs(tj).copy(tj).cmul(go)
+        if (dim == 1) iInc.resize(Array(iInc.size(2)))
+        gradInput.get[Tensor[T]](i) match {
+          case None => gradInput.update(i, iInc)
+          case Some(v) => v.add(iInc)
+        }
+
+        cc += 1
+      }
+    }
+
+    gradInput
+  }
+
+  protected def checkEmbeddingSize(t: Tensor[T]): Unit = {
+    val size = if (t.dim() == 1) t.size(1) else t.size(2)
+    require(embeddingSize <= 0 || embeddingSize == size,
+      s"size of input Tensor($size) not equal to embeddingSize($embeddingSize)!")
+  }
+
+  protected def dot(t1: Tensor[T], t2: Tensor[T]): Tensor[T] = {
+    var (input1, input2) = (t1, t2)
```

Comment:
why not use the ```tensor.dot``` method?

Reply:
@yiheng  We do batch dot operations for dim(2) here, which Tensor.dot don't support.